internal/data/trace.go

```diff
@@ -0,0 +1,377 @@
+}
+
+func (m *Span) SetEndTime(v TimestampUnixNano) {
+	m.orig.EndTimeUnixnano = uint64(v)
+}
+
+func (m *Span) SetAttributes(v Attributes) {
+	m.attributes = v.attrs
+	m.orig.DroppedAttributesCount = v.droppedCount
+}
+
+func (m *Span) SetEvents(v []*SpanEvent) {
+	m.events = v
+}
+
+func (m *Span) SetDroppedEventsCount(v uint32) {
+	m.orig.DroppedEventsCount = v
+}
+
+func (m *Span) SetLinks(v []*SpanLink) {
+	m.links = v
+}
+
+func (m *Span) SetDroppedLinksCount(v uint32) {
+	m.orig.DroppedLinksCount = v
+}
+
+func (m *Span) SetStatus(v SpanStatus) {
+	m.orig.Status = v.orig
+}
+
+type SpanStatus struct {
+	orig *otlptrace.Status
+}
+
+func NewSpanStatus(code otlptrace.Status_StatusCode, message string) SpanStatus {
+	return SpanStatus{orig: &otlptrace.Status{
+		Code:    code,
+		Message: message,
+	}}
+}
+
+// SpanEvent is a time-stamped annotation of the span, consisting of user-supplied
+// text description and key-value pairs.
+// Must use NewSpanEvent* function to create new instances.
+// Important: zero-initialized instance is not valid for use.
+type SpanEvent struct {
+	// Wrap OTLP Event.
+	orig *otlptrace.Span_Event
+
+	// Override attributes. This field is the source of truth for attributes.
+	// The counterpart stored in corresponding field of "orig" is ignored.
+	attributes AttributesMap
+}
+
+func NewSpanEvent(timestamp TimestampUnixNano, name string, attributes Attributes) *SpanEvent {
+	return &SpanEvent{
+		orig: &otlptrace.Span_Event{
+			TimeUnixnano:           uint64(timestamp),
+			Name:                   name,
+			DroppedAttributesCount: attributes.droppedCount,
+		},
+		attributes: attributes.attrs,
+	}
+}
+
+// NewSpanEventSlice creates a slice of SpanEvents that are correctly initialized.
+func NewSpanEventSlice(len int) []SpanEvent {
+	origs := make([]otlptrace.Span_Event, len)
+	wrappers := make([]SpanEvent, len)
+	for i := range origs {
+		wrappers[i].orig = &origs[i]
+	}
+	return wrappers
+}
+
+func (m *SpanEvent) Timestamp() TimestampUnixNano {
+	return TimestampUnixNano(m.orig.TimeUnixnano)
+}
+
+func (m *SpanEvent) Name() string {
+	return m.orig.Name
+}
+
+func (m *SpanEvent) Attributes() AttributesMap {
+	return m.attributes
+}
+func (m *SpanEvent) DroppedAttributesCount() uint32 {
+	return m.orig.DroppedAttributesCount
+}
+
+func (m *SpanEvent) SetTimestamp(v TimestampUnixNano) {
+	m.orig.TimeUnixnano = uint64(v)
+}
+
+func (m *SpanEvent) SetName(v string) {
+	m.orig.Name = v
+}
+
+func (m *SpanEvent) SetAttributes(v Attributes) {
+	m.attributes = v.attrs
+	m.orig.DroppedAttributesCount = v.droppedCount
+}
+
+// SpanLink is a pointer from the current span to another span in the same trace or in a
+// different trace. For example, this can be used in batching operations,
+// where a single batch handler processes multiple requests from different
+// traces or when the handler receives a request from a different project.
+// Must use NewSpanLink* function to create new instances.
```

Comment:
This comment should be on the other structs too.

Reply:
I think I added to all relevant structs. Do you see missing anywhere?

Reply:
Hum I thought that the intention was to have all structs here via New* but I see that some of them can work with their zero value.

Reply:
Yes, some can work via default values, but I think you have a good point. For uniformness and for being able to change to different representation it may be better to require New* for all structs.

Reply:
I added the same comment to Resource and to ResourceSpans.